[GeoMechanicsApplication] Adapt linear elastic laws

rfaasse opened the pull request:
**📝 Description**
Added the option to force compatibility with Plaxis's behavior when it comes to the K0 procedure. When compatibility is requested, all coupling terms in the elasticity matrix are set to zero. When the K0 procedure ends, the coupling terms are restored.

avdg81 committed:
Added a new (still empty) top-level base class

The new top-level base class is intended to provide common behavior for
all linear elastic constitutive laws.

avdg81 committed:
Added functions for serializing class `GeoLinearElasticLaw`

avdg81 committed:
The linear elastic law does not initialize its material response

This behavior has been moved from class `ElasticIsotropicK03DLaw` to
class `GeoLinearElasticLaw`. Furthermore, rather than providing an empty
implementation for `InitializeMaterialResponseCauchy`, we use member
`RequiresInitializeMaterialResponse` to skip the initialization.

Since class `GeoLinearElasticPlaneStrain2DLaw` did provide an override
to initialize the material response, we had to override member
`RequiresInitializeMaterialResponse` for this class, too.

avdg81 committed:
Moved `GetStrainMeasure` to the new base class

It is no longer overridden by class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved `GetStressMeasure` to the new base class

It is no longer overridden by class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Let `CalculatePK2Stress` start as a pure virtual member function

Any derived classes must provide an implementation for it.

avdg81 committed:
Let `CalculateElasticMatrix` start as a pure virtual member function

Any derived classes must provide an implementation for it.

avdg81 committed:
Let `CalculateCauchyGreenStrain` start as a pure virtual member function

Any derived classes must provide an implementation for it.

avdg81 committed:
Moved `CalculateMaterialResponsePK2` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved `CalculateMaterialResponsePK1` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved `CalculateMaterialResponseKirchhoff` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved `CalculateMaterialResponseCauchy` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved `RequiresFinalizeMaterialResponse` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved one overload of `CalculateValue` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved another overload of `CalculateValue` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved yet another overload of `CalculateValue` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved an overload of `SetValue` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Moved another overload of `SetValue` to the new base class

It has been removed from class `ElasticIsotropicK03DLaw`.

avdg81 committed:
Eliminated member `CheckClearElasticMatrix`

Its usages have been replaced by assigning a zero matrix with the
appropriate dimensions.

avdg81 committed:
Moved member `Check` to the new base class

avdg81 committed:
Removed an unused alias

avdg81 committed:
Moved one `using` statement to where it made sense

Also removed a `using` statement that was not needed at all.

avdg81 committed:
Changed the base class of `GeoLinearElasticPlaneStress2DLaw`

avdg81 committed:
Removed an unused alias

avdg81 committed:
Changed the base class of `LinearPlaneStrainK0Law`

avdg81 committed:
Avoid that inherited member functions are hidden

avdg81 committed:
Made several overridden member functions `public`

The access specifiers are now consistent with those from the base class.

Also made sure that the material finalization functions will be called.

avdg81 committed:
Changed base class in `using` statement

avdg81 committed:
Added a flag to control the usage of coupling terms in elastic laws

- When the process for the K0 procedure is initialized, make sure that
  the elasticity matrix does not contain any coupling terms. In this
  way, we can mimic Plaxis' K0 behavior.
- When the process for the K0 procedure is finalized, make sure to
  revert the elasticity matrix back to contain coupling terms.

avdg81 committed:
Added a first test for the adapted K0 procedure

The test comprises three horizontal soil layers. The test files have
been taken from Jon's comparison with Plaxis' K0 procedure.

avdg81 committed:
Added a second test for the adapted K0 procedure

The test comprises three tilted soil layers. The test files have been
taken from Jon's comparison with Plaxis' K0 procedure.

Also renamed a test checker function.

avdg81 committed:
Added a unit test for the apply K0 procedure

The new test verifies whether coupling terms are considered when the new
flag `force_plaxis_compatibility` has not been supplied.

To make this work, we have added a data member to class
`ApplyK0ProcedureProcess` which keeps the settings of the K0 procedure.

Also added a stub class for a linear elastic law.

avdg81 committed:
Added two more unit tests for forcing Plaxis compatibility

Also added a few helper functions to make it easier to read the (test)
code.

Furthermore, removed two comments that were no longer needed, since the
code has become more self-explaining.

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.cpp:
```diff
@@ -0,0 +1,202 @@
+    return rValue;
+}
+
+void GeoLinearElasticLaw::SetValue(const Variable<double>&, const double&, const ProcessInfo&)
+{
+    // Since `ConstitutiveLaw::SetValue` always throws an exception, an override
+    // is required when that is undesired behavior
+}
+
+void GeoLinearElasticLaw::SetValue(const Variable<Vector>&, const Vector&, const ProcessInfo&)
+{
+    // Since `ConstitutiveLaw::SetValue` always throws an exception, an override
+    // is required when that is undesired behavior
+}
```
These functions were not there in e.g. the k0 3d law, why are they added here?

avdg81 replied:
I think they were: see lines 216 to 222 of file `applications/GeoMechanicsApplication/custom_constitutive/elastic_isotropic_K0_3d_law.h`. Note that the (empty) function definitions were in the header file, not in the implementation file. Or am I overlooking something?

rfaasse replied:
Nope, I overlooked them, my bad (missed them in the cpp and didn't check for an implementation in the header)

rfaasse commented on applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.h:
```diff
@@ -0,0 +1,75 @@
+                                    ConstitutiveLaw::Parameters& rValues) = 0;
+
+private:
+    friend class Serializer;
+    void save(Serializer& rSerializer) const override;
+    void load(Serializer& rSerializer) override;
```
We are missing the save/load for the member variable.

avdg81 replied:
After our discussion, we've concluded that there is no need to serialize data member `mIsCouplingWanted`. The reason is this data member is supposed to be changed only during the execution of a stage that includes the K0 procedure. The serialization, however, aims at saving/loading the model state at the begin (or end) of a stage, but not during a stage run. When the need arises to serialize `mIsCouplingWanted`, we will add it.

avdg81 committed:
Several minor improvements

- Renamed a few functions, an `enum class`, and several variables.
- Use `BaseType` rather than repeating the base class name.
- Reformatted some code using clang-format.
- Added a comment and corrected another one.
- Moved two function definitions to a different location in the same
  file.
- Added a missing `#pragma once`.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_k0_procedure_process.cpp:
```diff
@@ -0,0 +1,105 @@
+    ApplyK0ProcedureProcess process{r_model_part, k0_settings};
+    process.ExecuteInitialize();
+
+    KRATOS_EXPECT_TRUE(AllElementsHaveSameCouplingOption(r_model_part.Elements(), GeoLinearElasticLaw::Coupling::No))
```
Maybe we can add the finalize of the process here as well and check that coupling is set to yes again

avdg81 replied:
Good suggestion! I have added a separate unit test to cover that behavior.

rfaasse commented on applications/GeoMechanicsApplication/tests/test_k0_procedure_process.py:
```diff
@@ -8,6 +8,24 @@ class KratosGeoMechanicsK0ProcedureProcessTests(KratosUnittest.TestCase):
+    def assert_stresses_at_integration_point(self, cauchy_stress_tensors, integration_point, expected_horizontal_stress, expected_vertical_stress, rel_tol):
+        """
+        Verifies whether the computed stresses are (nearly) equal to some expected values at the
+        given integration point.  Note that this function assumes there are no shear stresses!
```
This might be a bit more clear (same with the other tests)
```suggestion
        given integration point.  Note that this test assumes there are no shear stresses!
```

avdg81 replied:
What I tried to express here is that function `assert_stresses_at_integration_point` is intended to check the stress state after applying the K0 procedure. In other words, it cannot be used to check _any_ stress state, but only the ones produced by the K0 procedure (which, by definition, will yield shear stresses that are equal to zero). From that point of view, it is not the test which assumes there are no shear stresses, but this check function. The reason for this comment is to avoid that this check function will be accidentally reused to verify stress states that were not produced by the K0 procedure. Therefore, I think we should not replace "function" by "test". We could, however, replace "function" by "checker" or "assertion". Would that be clearer in your opinion?

rfaasse replied:
Okay clear, in that case indeed the current comment is as it should be! I misread and thought this was a test as well, but indeed I now see this is a helper function

avdg81 committed:
Attempted to fix some issues found by Codacy

All of the reported issue were related to style.

rfaasse reviewed (commented):
Looks good, just a few comments from my side.

avdg81 committed:
A few more minor improvements

- Added two comments that explain why we don't serialize data member
  `mIsCouplingWanted` of class `GeoLinearElasticLaw`.
- Added a unit test that verifies whether a change to the coupling
  option is in effect during the execution of the K0 procedure only.

avdg81 committed:
Second attempt to fix the indentation issues found by Codacy

The applied fix is based on the reply found here: https://github.com/remarkjs/remark-lint/issues/178#issuecomment-476716698

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.cpp:
```diff
@@ -0,0 +1,202 @@
+void GeoLinearElasticLaw::CalculateMaterialResponsePK2(ConstitutiveLaw::Parameters& rValues)
+{
+    KRATOS_TRY
+    // b.- Get Values to compute the constitutive law:
```
What is the intention of b.- here?

avdg81 replied:
I've moved several functions verbatim from one file to this one. To not complicate the code review (i.e. to make clear this code was only moved), I was reluctant to make any other changes. (Those can be done in a separate PR.) For improving comments and addressing code smells, however, we can make an exception and fix them right away. So I will do that here. Thanks for pointing this one out.

WPK4FEM replied:
done

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.cpp:
```diff
@@ -0,0 +1,202 @@
+    Vector& r_strain_vector = rValues.GetStrainVector();
+
+    // NOTE: SINCE THE ELEMENT IS IN SMALL STRAINS WE CAN USE ANY STRAIN MEASURE. HERE EMPLOYING THE CAUCHY_GREEN
+    if (r_options.IsNot(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN)) {
+        CalculateCauchyGreenStrain(rValues, r_strain_vector);
+    }
+
+    if (r_options.Is(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR)) {
+        Matrix& r_constitutive_matrix = rValues.GetConstitutiveMatrix();
+        CalculateElasticMatrix(r_constitutive_matrix, rValues);
+    }
+
+    if (r_options.Is(ConstitutiveLaw::COMPUTE_STRESS)) {
+        Vector& r_stress_vector = rValues.GetStressVector();
+        CalculatePK2Stress(r_strain_vector, r_stress_vector, rValues);
+    }
```
rValues is the output argument here. That raises the expectation that the if conditions should be mutually exclusive, otherwise rValues contains the last computed content.

avdg81 replied:
Yes, I agree with your suggestion to make them mutually exclusive. However, it turned out that we have misunderstood this part of the code. Basically, it represents the three steps to calculate the stress state at an integration point: (1) calculate the Green strain from the given strain measure (whatever that is), (2) calculate the constitutive matrix, and (3) calculate the stress tensor from the Green strain and the constitutive matrix. Each of these steps appears to be optional, which implies that some cached value is used when a calculation is omitted. This code can (and probably should) be made clearer, but that goes beyond the scope of this PR.

WPK4FEM replied:
We checked together and these are not mutually exclusive.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.cpp:
```diff
@@ -0,0 +1,202 @@
+    KRATOS_ERROR_IF_NOT(rMaterialProperties.Has(POISSON_RATIO))
+        << "POISSON_RATIO is not available in material parameters" << std::endl;
+
+    const double& nu = rMaterialProperties[POISSON_RATIO];
+    KRATOS_ERROR_IF((nu > 0.499 && nu < 0.501) || (nu < -0.999 && nu > -1.01))
+        << "POISSON_RATIO has invalid value " << std::endl;
```
strange range, I would expect -1.0 <= nu < 0.5 ( give or take equal signs )

avdg81 replied:
Good catch! Definitely something that needs to be addressed, but not in this PR. I will add an issue to our product backlog to make sure we don't forget about it.

WPK4FEM replied:
Out of scope here, there now is an issue on the product backlog.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_k0_procedure_process.cpp:
```diff
@@ -22,18 +22,51 @@
+            dynamic_cast<GeoLinearElasticLaw*>(rElement.GetProperties().GetValue(CONSTITUTIVE_LAW).get());
+        if (pLinearElasticLaw) {
+            pLinearElasticLaw->SetCouplingOption(WantCoupling);
+        }
```
What does the defined clang format say here? I expected a 1 line if statement, but maybe it exceeds 100 characters.

avdg81 replied:
Done. (It was not done automatically by clang-format; but after modifying it manually, it was also not reverted by clang-format.)

WPK4FEM reviewed (commented):
Nice work including some unraveling of the class structure. 
Good that it mentions coupling. The opposite of it is decoupling. I wonder if we should avoid explicit mentioning of Plaxis throughout everything here.

markelov208 commented on applications/GeoMechanicsApplication/custom_constitutive/linear_elastic_law.cpp:
```diff
@@ -0,0 +1,202 @@
+
+void GeoLinearElasticLaw::CalculateMaterialResponseKirchhoff(ConstitutiveLaw::Parameters& rValues)
+{
+    CalculateMaterialResponsePK2(rValues);
```
A general question, most likely for a separate issue, here CalculateMaterialResponseKirchhoff calls CalculateMaterialResponsePK2. Below CalculateMaterialResponseCauchy calls CalculateMaterialResponsePK2. When SmallStrainUDSM3DLaw::CalculateMaterialResponsePK2 calls CalculateMaterialResponseCauchy. It looks that in both cases function names do not reflect the numerics and here CalculateMaterialResponsePK2 should be split in appropriate functions.

avdg81 replied:
Thanks for pointing out this inconsistency, Gennady! I agree with you that this can be improved. But, as you indicated yourself, this is beyond the scope of this PR. I have created an issue for it on our product backlog.

markelov208 reviewed (commented):
Anne, sorry only one general question.

avdg81 committed:
Processed some review feedback

- Removed an `#include` that was copied from an implementation file to a
  header file.
- Short `if` statements can be put on a single line.
- Renamed setting `"force_plaxis_compatibility"` to
  `"force_decoupled_elasticity"`.
- Removed a comment that had no added value.
- Renamed a few unit test cases.

avdg81 committed:
Renamed two member functions and replaced a data member

- Renamed member functions `GetCouplingOption` and `SetCouplingOption`
  of class `GeoLinearElasticLaw` to
  `GetConsiderDiagonalEntriesOnlyAndNoShear` and
  `SetConsiderDiagonalEntriesOnlyAndNoShear`, respectively.
- The type of the corresponding data member has been changed from
  `IsCouplingWanted` (an `enum class`) to `bool`. As a result, type
  `IsCouplingWanted` could be removed.
- Renamed a helper function.

avdg81 committed:
Renamed the flag to use the "standard" K0 procedure

It used to be named "force_decoupled_elasticity". Now the flag is named
"use_standard_procedure".

Also renamed the function that retrieves the flag's value.

Note that the behavior when the flag is not defined has been changed. In
that case, we now use the "standard" procedure, i.e. consider the
diagonal entries only and set the shear terms to zero.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_k0_procedure_process.cpp:
```diff
@@ -47,75 +50,72 @@ ModelPart& PrepareTestModelPart(Model& rModel)
 
-    KRATOS_EXPECT_TRUE(AllElementsHaveSameCouplingOption(
-        r_model_part.Elements(), GeoLinearElasticLaw::IsCouplingWanted::Yes))
+    KRATOS_EXPECT_TRUE(boost::algorithm::all_of(r_model_part.Elements(), ElementConsidersDiagonalEntriesOnlyAndNoShear))
```
Is there a specific reason we move to the boost version instead of the stl one with iterators? I like the readability, just wondering about stl vs 'third party' (although boost is of course a special kind of 3rd party in my opinion)

avdg81 replied:
The primary reason to use `boost::algorithm::all_of` is it provides an overload that accepts a range (which I feel is easier to read than a pair of iterators; also, you are very explicit that you want to visit all elements in the container, and you can't accidentally pass a mismatching pair of iterators). As of C++20, ranges are available in the standard. Since we are still at C++17, this is a nice alternative. And since it doesn't pull in a new dependency, I feel this is acceptable. We may consider to use it in other places, too.

rfaasse commented on applications/GeoMechanicsApplication/custom_processes/apply_k0_procedure_process.cpp:
```diff
@@ -52,15 +52,15 @@ ApplyK0ProcedureProcess::ApplyK0ProcedureProcess(ModelPart& model_part, const Pa
-    if (ForceDecoupledElasticity()) {
-        SetCouplingOption(mrModelPart.Elements(), GeoLinearElasticLaw::IsCouplingWanted::No);
+    if (UseStandardProcedure()) {
+        SetConsiderDiagonalEntriesOnlyAndNoShear(mrModelPart.Elements(), true);
```
I think this reads very clearly now

avdg81 replied:
Thank you :-)

rfaasse commented on applications/GeoMechanicsApplication/tests/test_k0_procedure_process/README.md:
```diff
@@ -1,12 +1,13 @@
-- [K<sub>0</sub> procedure with umat](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_umat/README.md)
-- [K<sub>0</sub> procedure for simple dike](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_simple_dike/README.md)
-
+-   [K<sub>0</sub> procedure normal consolidation](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_nc/README.md)
+-   [K<sub>0</sub> procedure normal consolidation with layers](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_nc_layers/README.md)
+-   [K<sub>0</sub> procedure normal consolidation with OCR](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_nc_ocr/README.md)
+-   [K<sub>0</sub> procedure normal consolidation with skew layers](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_nc_skew_layers/README.md)
+-   [K<sub>0</sub> procedure normal consolidation with skew layers and dam](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_nc_skew_layers_dam/README.md)
+-   [K<sub>0</sub> procedure with umat](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_k0_umat/README.md)
+-   [K<sub>0</sub> procedure for simple dike](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_simple_dike/README.md)
+-   [K<sub>0</sub> procedure for horizontal layers](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_with_horizontal_layers/README.md)
+-   [K<sub>0</sub> procedure for tilted layers](https://github.com/KratosMultiphysics/Kratos/tree/master/applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_with_tilted_layers/README.md)
```
Shouldn't we make changes to all the k0 tests which were there before to explicitly set it the 'use-standard-procedure' to false? Or do all tests pass now?

avdg81 replied:
Since all tests pass, I decided to not add the "use_standard_procedure" flag to the existing test files. But I will double-check with @WPK4FEM to make sure that I did not make any mistake here.

WPK4FEM replied:
Anne has checked today.

rfaasse reviewed (commented):
I think the changes look clear, a few final questions.

Thanks for the work and rework on this PR!

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_k0_procedure_process.cpp:
```diff
@@ -22,16 +22,46 @@
+
+using namespace Kratos;
+
+void SetConsiderDiagonalEntriesOnlyAndNoShear(ModelPart::ElementsContainerType& rElements, bool Whether)
```
I think the AndNoShear referred to the last diagonal entries. Those are now left intact, thus making this a less descriptive name.

WPK4FEM replied:
Mistake of Wijtze Pieter, checked again in the code and the G term on the (XY, XY) diagonal really becomes 0. Ignore the remark.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_processes/apply_k0_procedure_process.cpp:
```diff
@@ -22,16 +22,46 @@
+
+void ApplyK0ProcedureProcess::ExecuteFinalize()
+{
+    if (UseStandardProcedure()) {
```
Does clang format not take out the { } for this 1 line like statement?

avdg81 replied:
No, it does not remove the curly braces automatically. But it also does not put them back when I take them out manually. And so I did.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_apply_k0_procedure_process.cpp:
```diff
@@ -0,0 +1,121 @@
+    Parameters k0_settings{R"({"use_standard_procedure": true})"};
+
+    ApplyK0ProcedureProcess process{r_model_part, k0_settings};
+    process.ExecuteInitialize(); // start considering diagonal entries only and no shear
+    process.ExecuteFinalize(); // stop considering diagonal entries only and no shear
```
Remaining "and no shear" remark, I think this should be removed.

WPK4FEM replied:
See previous remark. The (XY,XY) diagonal term really becomes 0.

WPK4FEM commented on applications/GeoMechanicsApplication/tests/test_k0_procedure_process/test_k0_procedure_with_tilted_layers/ProjectParameters.json:
```diff
@@ -0,0 +1,246 @@
+        }]
+    },
+    "processes": {
+        "auxilliary_process_list": [
```
I still am confused. Apparently both auxilliary_process_list and auxilliar_process_list both work, but I do not understand how and why.

WPK4FEM replied:
This is out of scope here, we have made a ticket in the product backlog.

avdg81 committed:
Extended an existing test for the K0 procedure

It uses the non-standard procedure and it checks the horizontal stress
components at three integration points.

avdg81 committed:
Removed redundant curly braces

WPK4FEM reviewed (approved):
Done.
